[AMDGPU][NPM] Port AMDGPUArgumentUsageInfo to NPM

PrasoonMishra opened the pull request:
Port AMDGPUArgumentUsageInfo analysis to the NPM to fix suboptimal code generation when NPM is enabled by default.

Previously, DAG.getPass() returns nullptr when using NPM, causing the argument usage info to be unavailable during ISel. This resulted in fallback to FixedABIFunctionInfo which assumes all implicit arguments are needed, generating unnecessary register setup code for entry functions.

Fixes LLVM::CodeGen/AMDGPU/cc-entry.ll

Changes:
- Split AMDGPUArgumentUsageInfo into a data class and NPM analysis wrapper
- Update SIISelLowering to use DAG.getMFAM() for NPM path
- Add RequireAnalysisPass in addPreISel() to ensure analysis availability

This follows the same pattern used for PhysicalRegisterUsageInfo.

PrasoonMishra committed:
[AMDGPU][NPM] Port AMDGPUArgumentUsageInfo to NPM

Port AMDGPUArgumentUsageInfo analysis to the NPM to fix suboptimal
code generation when NPM is enabled by default.

Previously, DAG.getPass() returns nullptr when using NPM, causing the
argument usage info to be unavailable during ISel. This resulted in
fallback to FixedABIFunctionInfo which assumes all implicit arguments
are needed, generating unnecessary register setup code for entry
functions.

Fixes LLVM::CodeGen/AMDGPU/cc-entry.ll

Changes:
- Split AMDGPUArgumentUsageInfo into a data class and NPM analysis wrapper
- Update SIISelLowering to use DAG.getMFAM() for NPM path
- Add RequireAnalysisPass in addPreISel() to ensure analysis availability

This follows the same pattern used for PhysicalRegisterUsageInfo.

PrasoonMishra committed:
Apply clang-format.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUArgumentUsageInfo.cpp:
```diff
@@ -191,3 +197,11 @@ AMDGPUArgumentUsageInfo::lookupFuncArgInfo(const Function &F) const {
   return I->second;
 }
+
+AnalysisKey AMDGPUArgumentUsageAnalysis::Key;
```
```suggestion
AnalysisKey AMDGPUArgumentUsageAnalysis::Key;

```

arsenm reviewed (commented):
This resulted in fallback to FixedABIFunctionInfo which assumes all implicit arguments are needed, generating unnecessary register setup code for entry functions.

PrasoonMishra committed:
Clean up data class interface

Fix:
- Removed legacy PM method names (doInitialization/doFinalization)
from AMDGPUArgumentUsageInfo data class.
- Simplified NPM run() as no special initialization is needed.

arsenm reviewed (approved):
LGTM but we should probably eliminate this pass

PrasoonMishra commented:
> LGTM but we should probably eliminate this pass

I looked into eliminating this pass. It doesn't seem straightforward as I can't use MMI to look up the callee's MF (i.e. passSpecialInputs fn needs the callee's ArgInfo during the caller's ISel) as FreeMachineFunction deletes each function's MF from MMI after its codegen completes. So by the time the caller is being lowered, the callee's MF may already be destroyed (or not yet created). I also thought of reconstructing ArgInfo from IR which is not quite trivial for entry functions.

I feel the current way of having scratchpad that stores ArgInfo is not too bad or expensive. What are your thoughts on this @arsenm?
